src/data/parquet_parser.h

```diff
@@ -0,0 +1,189 @@
+ * \author Chengyang Gu
+ */
+#ifndef DMLC_DATA_PARQUET_PARSER_H_
+#define DMLC_DATA_PARQUET_PARSER_H_
+
+#include <dmlc/data.h>
+#include <dmlc/strtonum.h>
+#include <dmlc/parameter.h>
+#include <cmath>
+#include <cstring>
+#include <map>
+#include <string>
+#include <limits>
+#include <future>
+#include <algorithm>
+#include <memory>
+#include <vector>
+#include "../data/row_block.h"
+#include "../data/parser.h"
+#include "arrow/io/api.h"
+#include "parquet/api/reader.h"
+
+
+namespace dmlc {
+namespace data {
+
+struct ParquetParserParam : public Parameter<ParquetParserParam> {
+  std::string format;
+  int label_column;
+  int weight_column;
+  int nthreads;
+
+  DMLC_DECLARE_PARAMETER(ParquetParserParam) {
+    DMLC_DECLARE_FIELD(format).set_default("parquet")
+      .describe("File format.");
+    DMLC_DECLARE_FIELD(label_column).set_default(0)
+      .describe("Column index (0-based) that will put into label.");
+    DMLC_DECLARE_FIELD(weight_column).set_default(-1)
+      .describe("Column index that will put into instance weights.");
+    DMLC_DECLARE_FIELD(nthreads).set_default(2)
+      .describe("Column index that will put into instance weights.");
+  }
+};
+
+template <typename IndexType, typename DType = real_t>
+class ParquetParser : public ParserImpl<IndexType, DType> {
+ public:
+  ParquetParser(const std::string& filename,
+                const std::map<std::string, std::string>& args,
+                int nthread) : row_groups_read_(0), nthread_(nthread) {
+    param_.Init(args);
+    nthread_ = param_.nthreads;
+    CHECK_EQ(param_.format, "parquet");
+
+    parquet_reader_ = parquet::ParquetFileReader::OpenFile(filename, false);
+    metadata_ = parquet_reader_->metadata();
+    num_rows_ = metadata_->num_rows();
+    num_cols_ = metadata_->num_columns();
+    num_row_groups_ = metadata_->num_row_groups();
+
+    have_next_ = (num_rows_ != 0);
+  }
+
+  /*!
+   * \brief read in next several blocks of data
+   * \param data vector of data to be returned
+   * \return true if the data is loaded, false if reach end
+   */
+  virtual bool ParseNext(std::vector<RowBlockContainer<IndexType, DType> > *data);
+
+ protected:
+  virtual void ParseRowGroup(int row_group_id,
+                            RowBlockContainer<IndexType, DType> *out);
+
+  virtual size_t BytesRead(void) const {
+    return -1;
+  }
+
+  virtual void BeforeFirst(void) {}
+
+ private:
+  ParquetParserParam param_;
+  // handle for reading parquet files
+  std::unique_ptr<parquet::ParquetFileReader> parquet_reader_;
+  std::shared_ptr<parquet::FileMetaData> metadata_;
+  // number of rows having read
+  int num_rows_;
+  int num_cols_;
+  int num_row_groups_;
+  int row_groups_read_;
+  // whether we have reached end of parquet file
+  bool have_next_;
+  // number of threads; hardcoded 4 for now
+  int nthread_;
+  // OMPException object to catch and rethrow exceptions in omp blocks
+  dmlc::OMPException omp_exc_;
```

Comment:
I don't think you need this one if you are using async.

Reply:
`omp_exc_` is used here https://github.com/bridgream/dmlc-core/blob/parquet_support/src/data/parquet_parser.h#L119 but I'm also not sure whether `omp_exc_` is still needed. 

Or maybe the question is: do we intend to deprecate OpenMP tasks in favor of `std::async`?

Reply:
I don't have a preference.  The comment is just, if one uses OpenMP or plain thread, then the guard is needed.  If one uses async as in the current form of this PR, then we can remove the guard.  Pick one method and we call it a day. ;-)

Reply:
I've experimented and it looks the guard is not needed if no exception is thrown. Maybe we can get rid of `dmlc::OMPException` for now? I've committed a version without `omp_exc_` but can revert it easily as it's the last one.